b_cfn_elasticsearch_index/source/action.py

```diff
@@ -0,0 +1,308 @@
+import base64
+import json
+import logging
+import os
+from typing import Dict, Any, Optional, Tuple
+from secrets import token_urlsafe
+
+import boto3
+from elasticsearch import Elasticsearch, RequestsHttpConnection
+from elasticsearch.helpers import reindex as elasticsearch_reindex
+from requests_aws4auth import AWS4Auth
+
+
+logger = logging.getLogger(__name__)
+
+
+class Action:
+    def __init__(self, invocation_event: Dict[str, Any]):
+        self.__invocation_event: Dict[str, Any] = invocation_event
+        self.__parameters: Dict[str, Any] = invocation_event['ResourceProperties']
+        self.__old_parameters: Dict[str, Any] = invocation_event['OldResourceProperties']
+        self.__resource_id: Optional[str] = invocation_event.get('PhysicalResourceId')
+
+        try:
+            self.REGION = os.environ['AWS_REGION']
+
+            self.INDEX_PREFIX = self.__parameters['IndexPrefix']
+            self.ELASTICSEARCH_ENDPOINT = self.__parameters['ElasticsearchEndpoint']
+            self.INDEX_SETTINGS = self.__parameters['IndexSettings']
+            self.MAPPING_SETTINGS = self.__parameters['MappingSettings']
+            self.MIGRATE_DATA = self.__parameters['MigrateData']
+        except KeyError as ex:
+            logger.error(f'Missing environment: {repr(ex)}.')
+            raise
+        
+        if not isinstance(self.INDEX_SETTINGS, dict):
+            logger.error(f'eror') # TODO
+            raise ValueError(f'Error')
+        
+        if not isinstance(self.MAPPING_SETTINGS, dict):
+            logger.error(f'eror') # TODO
+            raise ValueError(f'Error')
+
+    def create(self) -> Tuple[Optional[Dict[Any, Any]], Optional[str]]:
+        """
+        Creates a resource.
+
+        :return: A tuple containing two items:
+            1. Custom data to return back to CloudFormation service.
+            2. Physical resource id (can be empty).
+        """
+        logger.info(f'Initiating resource creation with these parameters: {json.dumps(self.__parameters)}.')
+
+        index_name, physical_id = self.__create_index(
+            self.INDEX_PREFIX,
+            self.REGION,
+            self.ELASTICSEARCH_ENDPOINT,
+            self.INDEX_SETTINGS,
+            self.MAPPING_SETTINGS,
+        )
+
+        return {"IndexName": index_name}, physical_id
+
+
+    def update(self) -> Tuple[Optional[Dict[Any, Any]], Optional[str]]:
+        """
+        Updates a resource.
+
+        :return: A tuple containing two items:
+            1. Custom data to return back to CloudFormation service.
+            2. Physical resource id (can be empty).
+        """
+        logger.info(f'Initiating resource update with these parameters: {json.dumps(self.__parameters)}.')
+
+        try:
+            index_name = self.__resource_id.split("||")[1]
```

Comment:
I would like to see a nice comment of why you are splitting the resource id and taking the second parameter. I know the idea behind this, but it would be nice if it would be written as a comment.